File: android/src/main/kotlin/com/giphy/giphy_flutter_sdk/GiphyFlutterVideoManager.kt
package com.giphy.giphy_flutter_sdk

import android.content.Context
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel

class GiphyFlutterVideoManager : MethodChannel.MethodCallHandler {
    private lateinit var channel: MethodChannel
    private lateinit var context: Context

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "muteAll" -> muteAll(result)
            "pauseAll" -> pauseAll(result)
            "resume" -> resume(result)
            else -> result.notImplemented()
        }
    }

    private fun muteAll(result: MethodChannel.Result) {
        GiphyFlutterSharedVideoPlayer.mute()
        result.success(null)
    }

    private fun pauseAll(result: MethodChannel.Result) {
        GiphyFlutterSharedVideoPlayer.pause()
        result.success(null)
    }

    private fun resume(result: MethodChannel.Result) {
        GiphyFlutterSharedVideoPlayer.resume()
        result.success(null)
    }

    fun onAttachedToEngine(flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        context = flutterPluginBinding.applicationContext
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "com.giphyfluttersdk/videoManager")
        channel.setMethodCallHandler(this)
    }

    fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}

File: ios/Classes/GiphyFlutterVideoManager.swift
//
//  GiphyFlutterVideoManager.swift
//  giphy_flutter_sdk
//
//  Created by Alex on 25.3.2024.
//

import Foundation

import Flutter
import UIKit
import GiphyUISDK

class GiphyFlutterVideoManager: NSObject {
    
    fileprivate var channel: FlutterMethodChannel?
    
    public override init() {
        super.init()
    }
    
    deinit {
    }
    
    func onAttachedToEngine(flutterPluginBinding: FlutterPluginRegistrar) {
        self.channel = FlutterMethodChannel(name: "com.giphyfluttersdk/videoManager", binaryMessenger: flutterPluginBinding.messenger())
        channel?.setMethodCallHandler({
            [weak self] (call: FlutterMethodCall, result: @escaping FlutterResult) -> Void in
            guard let self = self else {
                result(FlutterError(code: "SELF_NIL", message: "Reference to self is nil", details: nil))
                return
            }            
            
            switch call.method {
            case "muteAll":
                DispatchQueue.main.async {
                    if GiphyFlutterSharedVideoPlayer.initialized {
                        GiphyFlutterSharedVideoPlayer.shared.mute(true)
                    }
                }
                result(nil)
            case "pauseAll":
                DispatchQueue.main.async {
                    if GiphyFlutterSharedVideoPlayer.initialized {
                        GiphyFlutterSharedVideoPlayer.shared.pause()
                    }
                }
                result(nil)
            case "resume":
                DispatchQueue.main.async {
                    if GiphyFlutterSharedVideoPlayer.initialized {
                        GiphyFlutterSharedVideoPlayer.shared.resume()
                    }
                }
                result(nil)
                
            default:
                result(FlutterMethodNotImplemented)
            }
        })

    }
    
    func onDettachedFromEngine(flutterPluginBinding: FlutterPluginRegistrar) {
        self.channel?.setMethodCallHandler(nil)
        self.channel = nil
    }
}

